Dalamud/Interface/Internal/DalamudIme.cs

```diff
@@ -0,0 +1,630 @@
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Text.Unicode;
+
+using Dalamud.Game.Text;
+using Dalamud.Hooking.WndProcHook;
+using Dalamud.Interface.GameFonts;
+using Dalamud.Interface.Utility;
+using Dalamud.Logging.Internal;
+
+using ImGuiNET;
+
+using TerraFX.Interop.Windows;
+
+using static TerraFX.Interop.Windows.Windows;
+
+namespace Dalamud.Interface.Internal;
+
+/// <summary>
+/// This class handles CJK IME.
+/// </summary>
+[ServiceManager.EarlyLoadedService]
+internal sealed unsafe class DalamudIme : IDisposable, IServiceType
+{
+    private static readonly ModuleLog Log = new("IME");
+
+    private static readonly UnicodeRange[] HanRange =
+    {
+        UnicodeRanges.CjkRadicalsSupplement,
+        UnicodeRanges.CjkSymbolsandPunctuation,
+        UnicodeRanges.CjkUnifiedIdeographsExtensionA,
+        UnicodeRanges.CjkUnifiedIdeographs,
+        UnicodeRanges.CjkCompatibilityIdeographs,
+        UnicodeRanges.CjkCompatibilityForms,
+        // No more; Extension B~ are outside BMP range
+    };
+
+    private static readonly UnicodeRange[] HangulRange =
+    {
+        UnicodeRanges.HangulJamo,
+        UnicodeRanges.HangulSyllables,
+        UnicodeRanges.HangulCompatibilityJamo,
+        UnicodeRanges.HangulJamoExtendedA,
+        UnicodeRanges.HangulJamoExtendedB,
+    };
+
+    private readonly ImGuiSetPlatformImeDataDelegate setPlatformImeDataDelegate;
+
+    [ServiceManager.ServiceConstructor]
+    private DalamudIme() => this.setPlatformImeDataDelegate = this.ImGuiSetPlatformImeData;
+
+    /// <summary>
+    /// Finalizes an instance of the <see cref="DalamudIme"/> class.
+    /// </summary>
+    ~DalamudIme() => this.ReleaseUnmanagedResources();
+
+    private delegate void ImGuiSetPlatformImeDataDelegate(ImGuiViewportPtr viewport, ImGuiPlatformImeDataPtr data);
+
+    /// <summary>
+    /// Gets a value indicating whether Han(Chinese) input has been detected.
+    /// </summary>
+    public bool EncounteredHan { get; private set; }
+
+    /// <summary>
+    /// Gets a value indicating whether Hangul(Korean) input has been detected.
+    /// </summary>
+    public bool EncounteredHangul { get; private set; }
+
+    /// <summary>
+    /// Gets a value indicating whether to display the cursor in input text. This also deals with blinking.
+    /// </summary>
+    internal static bool ShowCursorInInputText
+    {
+        get
+        {
+            if (!ImGuiHelpers.IsImGuiInitialized)
+                return true;
```

Comment:
Should this be false?

Reply:
It doesn't matter; it decides whether to display the blinking `|` on text input control, and those earlier if statements are there just to prevent null dereferencing, in case something accesses that property before ImGui gets initialized.